guides/javascript.md

```diff
@@ -0,0 +1,264 @@
+  recipientIsOwnedAccount={isRecipientOwnedAccount}
+  recipientName={toName}
+  recipientNickname={toNickname}
+  senderName={fromName}
+  senderAddress={fromAddress}
+/>
+```
+
+## Use `async`/`await` syntax over `.then`/`.catch`
+
+Asynchronous code written using `async`/`await` syntax looks less complex and more straightforward than code written using `.then`/`.catch`. Additionally, using `async`/`await` leads to [better stack traces in Node and Chromium](https://mathiasbynens.be/notes/async-stack-traces), both of which use the V8 JavaScript engine, because when an asynchronous operation is `await`ed, the engine will remember the function where the `await` occurred, which means that it can place that function on the stack trace (otherwise, using `.then`/`.catch`, it would get lost).
+
+🚫
+
+``` javascript
+function makeRequest() {
+  return fetch("https://google.com")
+    .then((response) => {
+      return response.json().then((json) => {
+        return json['page_views'];
+      })
+    })
+}
+```
+
+✅
+
+``` javascript
+async function makeRequest() {
+  const response = await fetch("https://google.com");
+  const json = await response.json();
+  return json['page_views'];
+}
+```
+
+## `await` promises before returning them
+
+An `async` function that returns a rejected promise created via another `async` function may disappear from the stack trace. This is solved by `await`ing the promise before returning it.
+
+<details><summary><b>Read more</b></summary>
+<br/>
+<p>If you save the following to a file (say, <code>/tmp/example.js</code>) and run it with <code>node</code>:</p>
+
+``` javascript
+async function foo() {
+  return bar();
+}
+
+async function bar() {
+  await Promise.resolve();
+  throw new Error('BEEP BEEP');
+}
+
+foo().catch(error => console.log(error.stack));
+```
+
+then you will see the following in the terminal (as of Node 18):
+
+```
+Error: BEEP BEEP
+    at bar (/private/tmp/example.js:7:9)
+```
+
+Notice how `foo` is completely missing from the stack trace!
+
+However, if you put an `await` before the call to `bar`:
+
+``` javascript
+async function foo() {
+  return await bar();
+}
+
+async function bar() {
+  await Promise.resolve();
+  throw new Error('BEEP BEEP');
+}
+
+foo().catch(error => console.log(error.stack));
+```
+
+you will now see it at the bottom of the stack trace:
+
+```
+Error: BEEP BEEP
+    at bar (/private/tmp/example.js:7:9)
+    at async foo (/private/tmp/example.js:2:10)
+```
+
+</details>
+
+🚫
+
+``` javascript
+async function makeRequest() {
+  const response = await fetch('https://some/url');
+  return response.json();
+}
+```
+
+✅
+
+``` javascript
+async function makeRequest() {
+  const response = await fetch('https://some/url');
+  return await response.json();
+}
+```
+
+> [!WARNING]\
+> `await`ing promises has a significant performance impact when transpiling to ES6. [**Read more**](https://arthur.place/the-cost-of-return-await).
+
+## Use Jest's mock functions instead of Sinon
```

Comment:
These two sections seem out-of-place in this article. They seem better suited to an article about unit tests.

Reply:
Yeah, I think I'd separated testing guidelines in the original Notion doc, so that makes sense. I'll extract it for now, and we can expand on it in a future PR.

Reply:
Extracted here: [aa8fa2f](https://github.com/MetaMask/contributor-docs/pull/11/commits/aa8fa2f9319ca1140e556aa8fe4c3c512e5e5049)